### ISSUE TITLE
[bug]: autocreation of extension for AD users fails on Undefined array key "outboundcid"

### ISSUE THREAD

**Issue by LightningManGTS** [FreePBX/issue-tracker#76]
### FreePBX Version

FreePBX 17

### Issue Description

Extensions fail to create on users due to outboundcid being undefined. There is no way to match the object in AD through the web ui.

This is Microsoft Directory Sync to a Server 2022 domain, AD level 2016

Starting Sync on directory 'Tom&#039;s Help Desk'...
2024-04-08_17:38:09 -
2024-04-08_17:38:09 -Updating All Users
2024-04-08_17:38:09 -   _Sanatized_
2024-04-08_17:38:09 -   Retrieving all users...
2024-04-08_17:38:09 -   Got 8 users
2024-04-08_17:38:09 -           Updating jmerrick
2024-04-08_17:38:09 -                   Creating pjsip Extension 101
        Undefined array key "outboundcid"
Finished

### Operating Environment

Freepbx 17, Asterisk 21, 

-------------------+------------+-----------------------------------+-------------+-----------+
| Module            | Version    | Status                            | License     | Signature |
+-------------------+------------+-----------------------------------+-------------+-----------+
| amd               | 17.0.1     | Enabled                           | GPLv3+      | Sangoma   |
| announcement      | 17.0.2.1   | Enabled                           | GPLv3+      | Sangoma   |
| areminder         | 17.0.3.9   | Enabled                           | Commercial  | Sangoma   |
| arimanager        | 17.0.1     | Enabled                           | GPLv3+      | Sangoma   |
| asterisk-cli      | 17.0.2     | Enabled                           | GPLv3+      | Sangoma   |
| asteriskinfo      | 17.0.1     | Enabled                           | GPLv3+      | Sangoma   |
| backup            | 17.0.5.14  | Enabled                           | GPLv3+      | Sangoma   |
| blacklist         | 17.0.1.2   | Enabled                           | GPLv3+      | Sangoma   |
| broadcast         | 17.0.1.6   | Enabled                           | Commercial  | Sangoma   |
| builtin           |            | Enabled                           |             | Unsigned  |
| bulkhandler       | 17.0.4     | Enabled                           | GPLv3+      | Sangoma   |
| calendar          | 17.0.4.6   | Enabled                           | GPLv3+      | Sangoma   |
| callaccounting    | 17.0.4     | Enabled                           | Commercial+ | Sangoma   |
| callback          | 17.0.2.1   | Enabled                           | GPLv3+      | Sangoma   |
| callerid          | 17.0.1     | Enabled                           | Commercial  | Sangoma   |
| callforward       | 17.0.1.3   | Enabled                           | AGPLv3+     | Sangoma   |
| calllimit         | 17.0.1.2   | Enabled                           | Commercial  | Sangoma   |
| callrecording     | 17.0.3.6   | Enabled                           | AGPLv3+     | Sangoma   |
| callwaiting       | 17.0.3.2   | Enabled                           | GPLv3+      | Sangoma   |
| cdr               | 17.0.4.13  | Enabled                           | GPLv3+      | Sangoma   |
| cel               | 17.0.2.7   | Enabled                           | GPLv3+      | Sangoma   |
| certman           | 17.0.3.7   | Enabled                           | AGPLv3+     | Sangoma   |
| cidlookup         | 17.0.1.1   | Enabled                           | GPLv3+      | Sangoma   |
| conferences       | 17.0.3.1   | Enabled                           | GPLv3+      | Sangoma   |
| conferencespro    | 17.0.1.6   | Enabled                           | Commercial  | Sangoma   |
| configedit        | 17.0.1.1   | Enabled                           | AGPLv3+     | Sangoma   |
| contactmanager    | 17.0.5.6   | Enabled                           | GPLv3+      | Sangoma   |
| core              | 17.0.9.33  | Enabled                           | GPLv3+      | Sangoma   |
| cos               | 17.0.1.1   | Enabled                           | Commercial  | Sangoma   |
| customappsreg     | 17.0.1     | Enabled                           | GPLv3+      | Sangoma   |
| dashboard         | 17.0.4.2   | Enabled                           | AGPLv3+     | Sangoma   |
| daynight          | 17.0.1.1   | Enabled                           | GPLv3+      | Sangoma   |
| dictate           | 17.0.1     | Enabled                           | GPLv3+      | Sangoma   |
| directory         | 17.0.1.1   | Enabled                           | GPLv3+      | Sangoma   |
| donotdisturb      | 17.0.2.1   | Enabled                           | GPLv3+      | Sangoma   |
| endpoint          | 17.0.1.53  | Enabled                           | Commercial  | Sangoma   |
| extensionroutes   | 17.0.1     | Enabled                           | Commercial  | Sangoma   |
| extensionsettings | 17.0.1     | Enabled                           | GPLv3+      | Sangoma   |
| fax               | 17.0.3.2   | Enabled                           | GPLv3+      | Sangoma   |
| faxpro            | 17.0.1.14  | Enabled                           | Commercial  | Sangoma   |
| featurecodeadmin  | 17.0.2     | Enabled                           | GPLv3+      | Sangoma   |
| filestore         | 17.0.2.9   | Enabled                           | AGPLv3      | Sangoma   |
| findmefollow      | 17.0.4.6   | Enabled                           | GPLv3+      | Sangoma   |
| firewall          | 17.0.1.13  | Enabled                           | AGPLv3+     | Sangoma   |
| framework         | 17.0.15.13 | Enabled                           | GPLv2+      | Sangoma   |
| hotelwakeup       | 17.0.1.5   | Enabled                           | GPLv2       | Sangoma   |
| iaxsettings       | 17.0.1     | Enabled                           | AGPLv3      | Sangoma   |
| infoservices      | 17.0.1     | Enabled                           | GPLv2+      | Sangoma   |
| ivr               | 17.0.5     | Enabled                           | GPLv3+      | Sangoma   |
| languages         | 17.0.1     | Enabled                           | GPLv3+      | Sangoma   |
| logfiles          | 17.0.2     | Enabled                           | GPLv3+      | Sangoma   |
| manager           | 17.0.3     | Enabled                           | GPLv2+      | Sangoma   |
| miscapps          | 17.0.3     | Enabled                           | GPLv3+      | Sangoma   |
| miscdests         | 17.0.1.1   | Enabled                           | GPLv3+      | Sangoma   |
| music             | 17.0.4     | Enabled                           | GPLv3+      | Sangoma   |
| oembranding       |            | Not Installed (Locally available) | Commercial  | Sangoma   |
| paging            | 17.0.3     | Enabled                           | GPLv3+      | Sangoma   |
| pagingpro         | 17.0.1.6   | Enabled                           | Commercial  | Sangoma   |
| parking           | 17.0.2     | Enabled                           | GPLv3+      | Sangoma   |
| parkpro           | 17.0.1.4   | Enabled                           | Commercial  | Sangoma   |
| pbxmfa            | 17.0.2     | Enabled                           | Commercial+ | Sangoma   |
| pinsets           | 17.0.3     | Enabled                           | GPLv3+      | Sangoma   |
| pinsetspro        | 17.0.2     | Enabled                           | Commercial  | Sangoma   |
| pm2               | 17.0.3.1   | Enabled                           | AGPLv3+     | Sangoma   |
| presencestate     | 17.0.2.1   | Enabled                           | GPLv3+      | Sangoma   |
| printextensions   | 17.0.1.2   | Enabled                           | GPLv3+      | Sangoma   |
| queueprio         | 17.0.1.4   | Enabled                           | GPLv3+      | Sangoma   |
| queues            | 17.0.1.7   | Enabled                           | GPLv2+      | Sangoma   |
| queuestats        | 17.0.1.6   | Enabled                           | Commercial  | Sangoma   |
| qxact_reports     | 17.0.2     | Enabled                           | Commercial  | Sangoma   |
| recording_report  | 17.0.3.6   | Enabled                           | Commercial  | Sangoma   |
| recordings        | 17.0.2.2   | Enabled                           | GPLv3+      | Sangoma   |
| restapps          | 17.0.1.14  | Enabled                           | Commercial  | Sangoma   |
| ringgroups        | 17.0.2.4   | Enabled                           | GPLv3+      | Sangoma   |
| sangomaconnect    | 17.0.1.20  | Enabled                           | Commercial  | Sangoma   |
| sangomartapi      | 17.0.1.14  | Enabled                           | Commercial  | Sangoma   |
| setcid            | 17.0.1.2   | Enabled                           | GPLv3+      | Sangoma   |
| sipsettings       | 17.0.6.3   | Enabled                           | AGPLv3+     | Sangoma   |
| sipstation        | 17.0.3.3   | Enabled                           | Commercial  | Sangoma   |
| sms               | 17.0.1.8   | Enabled                           | Commercial  | Sangoma   |
| smsconnector      |            | Not Installed (Locally available) | GPLv3+      | Unknown   |
| smsplus           | 17.0.3     | Enabled                           | Commercial  | Sangoma   |
| soundlang         | 17.0.4     | Enabled                           | GPLv3+      | Sangoma   |
| sysadmin          | 17.0.1.48  | Enabled                           | Commercial  | Sangoma   |
| timeconditions    | 17.0.1.13  | Enabled                           | GPLv3+      | Sangoma   |
| tts               | 17.0.1.1   | Enabled                           | GPLv3+      | Sangoma   |
| ttsengines        | 17.0.1     | Enabled                           | AGPLv3      | Sangoma   |
| ucp               | 17.0.4.13  | Enabled                           | AGPLv3+     | Sangoma   |
| userman           | 17.0.6.16  | Enabled                           | AGPLv3+     | Sangoma   |
| vmblast           | 17.0.1     | Enabled                           | GPLv3+      | Sangoma   |
| vmnotify          | 17.0.1.5   | Enabled                           | Commercial  | Sangoma   |
| voicemail         | 17.0.5.14  | Enabled                           | GPLv3+      | Sangoma   |
| voicemail_report  |            | Not Installed (Locally available) | Commercial  | Sangoma   |
| vqplus            | 17.0.1.13  | Enabled                           | Commercial  | Sangoma   |
| weakpasswords     | 17.0.1     | Enabled                           | GPLv3+      | Sangoma   |
| webrtc            | 17.0.2.1   | Enabled                           | GPLv3+      | Sangoma   |
+-------------------+------------+-----------------------------------+-------------+-----------+

**Comment by cdolese** [comment]
tested as well , i do see it failing to create ext 2004 but mine complains of the device id 

```
root@fpbx17:~# fwconsole userman --sync 2 --verbose --force
Starting Sync on directory 'SAD'...
2024-04-09_20:49:29 -
2024-04-09_20:49:29 -Updating All Users
2024-04-09_20:49:29 -   ldapsearch -w 123abc987 -H "ldap://my.ad.com:389" -D "aduser@my.ad.com" -b "dc=ad,dc=com" -s sub "(&(&(objectcategory=person)(samaccountname=*))(objectclass=user))"
2024-04-09_20:49:29 -   Retrieving all users...
2024-04-09_20:49:29 -   Got 11 users
2024-04-09_20:49:29 -           Updating Administrator
2024-04-09_20:49:29 -           Updating Guest
2024-04-09_20:49:29 -           Updating cdolese
2024-04-09_20:49:30 -                   Linking Extension 2001 to cdolese
2024-04-09_20:49:30 -           Updating krbtgt
2024-04-09_20:49:30 -           Updating myadsrv
2024-04-09_20:49:30 -           Updating userone
2024-04-09_20:49:30 -                   Linking Extension 2001 to userone
2024-04-09_20:49:30 -           Updating usertwo
2024-04-09_20:49:30 -                   Linking Extension 2002 to usertwo
2024-04-09_20:49:30 -           Updating userthree
2024-04-09_20:49:30 -                   Linking Extension 2003 to userthree
2024-04-09_20:49:30 -           Updating userfour
2024-04-09_20:49:30 -                   Creating pjsip Extension 2004
        This device id is already in use
Finished
```

**Comment by LightningManGTS** [comment]
> tested as well , i do see it failing to create ext 2004 but mine complains of the device id
> 
> ```
> root@fpbx17:~# fwconsole userman --sync 2 --verbose --force
> Starting Sync on directory 'SAD'...
> 2024-04-09_20:49:29 -
> 2024-04-09_20:49:29 -Updating All Users
> 2024-04-09_20:49:29 -   ldapsearch -w 123abc987 -H "ldap://my.ad.com:389" -D "aduser@my.ad.com" -b "dc=ad,dc=com" -s sub "(&(&(objectcategory=person)(samaccountname=*))(objectclass=user))"
> 2024-04-09_20:49:29 -   Retrieving all users...
> 2024-04-09_20:49:29 -   Got 11 users
> 2024-04-09_20:49:29 -           Updating Administrator
> 2024-04-09_20:49:29 -           Updating Guest
> 2024-04-09_20:49:29 -           Updating cdolese
> 2024-04-09_20:49:30 -                   Linking Extension 2001 to cdolese
> 2024-04-09_20:49:30 -           Updating krbtgt
> 2024-04-09_20:49:30 -           Updating myadsrv
> 2024-04-09_20:49:30 -           Updating userone
> 2024-04-09_20:49:30 -                   Linking Extension 2001 to userone
> 2024-04-09_20:49:30 -           Updating usertwo
> 2024-04-09_20:49:30 -                   Linking Extension 2002 to usertwo
> 2024-04-09_20:49:30 -           Updating userthree
> 2024-04-09_20:49:30 -                   Linking Extension 2003 to userthree
> 2024-04-09_20:49:30 -           Updating userfour
> 2024-04-09_20:49:30 -                   Creating pjsip Extension 2004
>         This device id is already in use
> Finished
> ```

When it fails to make the extension in freepbx, it still ends up making it as a record in the mysql tables. my "work around" is to delete said entries as follows (replace 101 with your extension) and set AD to NOT auto create the extensions and then just make them manually

mysql -u root -D asterisk -e "DELETE FROM devices WHERE id = '101';"
mysql -u root -D asterisk -e "DELETE FROM users WHERE extension = '101';"

TLDR my theory is autoboundcid is probably some matching record their adding, they haven't finished it yet, and then made it required by accident. Or its just required by accident.

**Comment by LightningManGTS** [comment]
While I'm in here, I get syntex errors when trying to sync groups, it ends up stopping on said error after syncing the first group and its members. 

Anyone know how to turn group syncing off in the meantime?

**Comment by jissphilip** [comment]
@LightningManGTS 
You  DB has some duplicate entries ,  
Please go to advancedsettings 
![image](https://github.com/FreePBX/issue-tracker/assets/30717824/da5bfd62-dcb0-4d7b-af4b-bde00f47c7ea)
Change the "User & Devices Mode"  to DeviceandUser 
Then 
Applications-> Users  ( you can see the duplicate entries ) 
Applications-> Device ( you can see the duplicate entries )
Please remove  all duplicate and try to sync again

**Comment by jissphilip** [comment]
core version 17.0.9.35

**Comment by LightningManGTS** [comment]
> @LightningManGTS You DB has some duplicate entries , Please go to advancedsettings ![image](https://private-user-images.githubusercontent.com/30717824/321897381-da5bfd62-dcb0-4d7b-af4b-bde00f47c7ea.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI5MTk4NTEsIm5iZiI6MTcxMjkxOTU1MSwicGF0aCI6Ii8zMDcxNzgyNC8zMjE4OTczODEtZGE1YmZkNjItZGNiMC00ZDdiLWFmNGItYmRlMDBmNDdjN2VhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDEyVDEwNTkxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdmZGJlOWYxNzQ5M2Q1NDY3NTE0MGQ4ZjRiNWY0NjU1YTA4N2QwMmIzODliOTlkMGNkMWY0MzgxZTY4YzI4NjImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.h71oCn9W59n_kUqxzLfUnzbv4cHa9mu1NXdxnLBYJ68) Change the "User & Devices Mode" to DeviceandUser Then Applications-> Users ( you can see the duplicate entries ) Applications-> Device ( you can see the duplicate entries ) Please remove all duplicate and try to sync again

I'm sorry did you not see where I verified I did not have duplicate entries? The sync errors on Undefined array key "outboundcid" which is when it then makes a non fully formed extension that has to be removed. Currently this make ext auto creation non functional. Then there is the issue with groups where its failing on some sort of Syntex error after syncing the first group. After syncing the first group and stoping, if you then try and view the group that also fails with said Syntex error.

I also tried the group sync option where it allows you make the group in freepbx and then it syncs back to ad, that's not functional as well. It allows you to make the group but it does not sync back.

**Comment by jissphilip** [comment]
@LightningManGTS  did you test with  **  Userman 17.0.6.17 **

**Comment by LightningManGTS** [comment]
> @LightningManGTS did you test with ** Userman 17.0.6.17 **

Updating to core .35 and that version of Userman appear to have worked! give me a bit to test groups.

**Comment by LightningManGTS** [comment]
> > @LightningManGTS did you test with ** Userman 17.0.6.17 **
> 
> Updating to core .35 and that version of Userman appear to have worked! give me a bit to test groups.

Groups seems to be also working as well. Although if I change the Group DN scope it does not automatically remove the already sync'd entries (that should be fixed as currently there is no way to delete read only items) 

Apologies I also read what manage group locally does wrong and assumed it would write back when it does not.

**Comment by LightningManGTS** [comment]
I guess I will close this for now and put the issue with sync groups not automatically clearing themselves in an improvement issue post.
